docs/src/diplomacy/adder_tutorial.md

```diff
@@ -0,0 +1,322 @@
+
+```scala mdoc:invisible
+import chipsalliance.rocketchip.config.{Config, Parameters}
+import chisel3._
+import chisel3.internal.sourceinfo.SourceInfo
+import chisel3.stage.ChiselStage
+import chisel3.util.random.FibonacciLFSR
+import freechips.rocketchip.diplomacy.{SimpleNodeImp, RenderedEdge, ValName, SourceNode,
+                                       NexusNode, SinkNode, LazyModule, LazyModuleImp}
+```
+
+### Parameter negotiation and passing
+
+We have a width parameter we would like our modules to share. Any module that
+wants to send or receive parameter information must have **one or more nodes**.
+
+Between each pair of nodes, there may be **one or more edges**. The edge does
+the negotiation work to ensure parameter agreement between the nodes it connects.
+
+Let's pass around our parameters like so:
+
+![adder diplomatic nodes](diagrams/adder_nodes.pdf)
+
+Because the graph diplomacy creates is **acyclic**, it is directional. The
+direction towards the sinks is **downward**, and the direction towards the
+sources is **upward**.
+
+#### Parameters
+
+In our example, we want all of our circuitry to agree upon a Scala `Int` to use
+as a wire width. For clarity, we will define distinct case classes for our
+parameters, although they all contain the same type.
+
+```scala mdoc
+case class UpwardParam(width: Int)
+case class DownwardParam(width: Int)
+case class EdgeParam(width: Int)
+```
+
+#### Node Implementation
+
+In our "node implementation", or `NodeImp`, we describe how parameters flow
+throughout our graph and how parameters are negotiated between nodes.
+
+We will use `SimpleNodeImp`, which performs the same parameter negotiation and
+passes the same bundles along an edge, regardless of whether the edge points
+into our out of the node.
+
+Our **edge parameter** (`E`) describes the data type that needs to be passed along
+the edges in our graph; in our case we only need to keep track of one `Int` that
+describes our width, which will live inside our `EdgeParam` class.
+
+We also specify a **bundle parameter** (`B`), which is the type of data that will
+resolve into parameterized hardware ports between our modules. For us, this is
+a Chisel `UInt` with the width described by our edge parameter (see the `bundle`
+function).
+
+```scala mdoc
+// PARAMETER TYPES:                       D              U            E          B
+object AdderNodeImp extends SimpleNodeImp[DownwardParam, UpwardParam, EdgeParam, UInt] {
+  def edge(pd: DownwardParam, pu: UpwardParam, p: Parameters, sourceInfo: SourceInfo) = {
+    if (pd.width < pu.width) EdgeParam(pd.width) else EdgeParam(pu.width)
+  }
+  def bundle(e: EdgeParam) = UInt(e.width.W)
+  def render(e: EdgeParam) = RenderedEdge("blue", s"width = ${e.width}")
+}
+```
+
+Notice that the `edge` function does the actual negotiation between nodes. In our
+case, it compares the parameters traveling both upwards and downwards, and chooses
+the one with the smaller width to pass along.
```

Comment:
I don't think we should say "pass along" here. Like, I would describe the downwards and upwards ones as "getting passed" since they actually move through the graph from node to node. But the edge ones are only visible to the two nodes connected to this specific edge. Like if you made your edges always pick `width = 3`, all the other parameters would still get passed around, they would just be ignored. Maybe `and chooses
the one with the smaller width as the final result to make visible from each node.`?